src/Lykke.AlgoStore.Job.Stopping/ExpiredInstancesMonitor.cs

```diff
@@ -0,0 +1,120 @@
+﻿using Common.Log;
+using Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Enumerators;
+using Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Models;
+using Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Repositories;
+using Lykke.AlgoStore.Job.Stopping.Settings.JobSettings;
+using Lykke.AlgoStore.KubernetesClient;
+using Lykke.AlgoStore.KubernetesClient.Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Lykke.AlgoStore.Job.Stopping
+{
+    public class ExpiredInstancesMonitor
+    {
+        private readonly IAlgoClientInstanceRepository _algoClientInstanceRepository;
+        private readonly IKubernetesApiClient _kubernetesApiClient;
+        private readonly ExpiredInstancesMonitorSettings _settings;
+        private const string _loggingContext = "Check for expired instances";
+        private readonly ILog _log;
+
+        public ExpiredInstancesMonitor(
+            IAlgoClientInstanceRepository algoClientInstanceRepository,
+            IKubernetesApiClient kubernetesApiClient,
+            ExpiredInstancesMonitorSettings settings,
+            ILog log)
+        {
+            _algoClientInstanceRepository = algoClientInstanceRepository;
+            _kubernetesApiClient = kubernetesApiClient;
+            _settings = settings;
+            _log = log;
+        }
+
+        public async Task StartAsync()
+        {
+            await ExecuteEvery();
+        }
+
+        private async Task ExecuteEvery()
+        {
+            await Task.Run(async () =>
+            {
+                while (true)
+                {
+                    var expiredInstances = await GetExpiredAlgoInstancesAsync();
+
+                    await TryStopExpiredInstances(expiredInstances);
+
+                    await Task.Delay(TimeSpan.FromSeconds(_settings.CheckIntervalInSeconds));
+                }
+            });
+        }
+
+        public async Task TryStopExpiredInstances(List<AlgoInstanceStoppingData> expiredInstances)
+        {
+            foreach (var instance in expiredInstances)
+            {
+                var instancePod = await GetInstancePodAsync(instance.InstanceId);
+                if (instancePod == null)
+                {
+                    await _log.WriteWarningAsync(nameof(ExpiredInstancesMonitor), _loggingContext, $"Instance {instance.InstanceId} of client id {instance.ClientId} is marked as started in db, but its pod was not found in kubernetis.");
+                    continue;
```

Comment:
Should we handle this case? Perhaps mark instance as stopped anyway?

Reply:
Yeah, we discussed it a lot with Dafi, at the end decided to to leave as it currently implemented (in API I think) - if pod is not found, dont change its state.